File: SmartReceipts/Model/TaxCalculator.m
//
//  TaxCalculator.m
//  SmartReceipts
//
//  Created by Jaanus Siim on 23/05/15.
//  Copyright (c) 2015 Will Baumann. All rights reserved.
//

#import "TaxCalculator.h"
#import "NSDecimalNumber+WBNumberParse.h"

@interface TaxCalculator ()

@property (nonatomic, strong) UITextField *priceField;
@property (nonatomic, strong) UITextField *taxField;
@property (nonatomic, strong) NSNumberFormatter *taxFormatter;

@end

@implementation TaxCalculator

- (id)initWithSourceField:(UITextField *)priceField targetField:(UITextField *)taxField {
    self = [super init];
    if (self) {
        _priceField = priceField;
        _taxField = taxField;
        [priceField addTarget:self action:@selector(priceChanged) forControlEvents:UIControlEventEditingChanged];
        _taxFormatter = [[NSNumberFormatter alloc] init];
        [_taxFormatter setNumberStyle:NSNumberFormatterCurrencyStyle];
        [_taxFormatter setCurrencySymbol:@""];
        [_taxFormatter setCurrencyGroupingSeparator:@""];
    }
    return self;
}

- (void)priceChanged {
    NSString *priceString = self.priceField.text;
    NSDecimalNumber *price = [NSDecimalNumber decimalNumberOrZeroUsingCurrentLocale:priceString];
    NSDecimalNumber *tax = [TaxCalculator taxWithPrice:price taxPercentage:self.taxPercentage priceEnteredPreTax:self.priceIsPreTax];
    [self.taxField setText:[self.taxFormatter stringFromNumber:tax]];
}

static NSDecimalNumber *__hundred;
+ (NSDecimalNumber *)hundred {
    if (!__hundred) {
        __hundred = [NSDecimalNumber decimalNumberWithString:@"100"];
    }

    return __hundred;
}

+ (NSDecimalNumber *)taxWithPrice:(NSDecimalNumber *)price taxPercentage:(NSDecimalNumber *)taxPercentage priceEnteredPreTax:(BOOL)preTax {
    NSDecimalNumber *tax;
    if (preTax) {
        tax = [[price decimalNumberByMultiplyingBy:taxPercentage] decimalNumberByDividingBy:self.hundred];
    } else {
        NSDecimalNumber *addeage = [self.hundred decimalNumberByAdding:taxPercentage];
        NSDecimalNumber *divided = [price decimalNumberByDividingBy:addeage];
        tax = [divided decimalNumberByMultiplyingBy:taxPercentage];
    }

    return tax;
}

@end

File: SmartReceipts/Model/TaxCalculator.swift
//
//  TaxCalculator.swift
//  SmartReceipts
//
//  Created by Bogdan Evsenev on 18/06/2017.
//  Copyright © 2017 Will Baumann. All rights reserved.
//

import RxSwift

class TaxCalculator: NSObject {
    
    let priceSubject = PublishSubject<Decimal?>()
    let taxSubject = PublishSubject<String>()
    var taxPercentage = NSDecimalNumber(decimal: 0)
    
    private let disposeBag = DisposeBag()
    private var priceIsPreTax = false
    private let formatter = NumberFormatter()
    
    fileprivate static let HUNDRED = NSDecimalNumber(decimal: 100)
    
    init(preTax: Bool) {
        super.init()
        priceIsPreTax = preTax
        taxPercentage = NSDecimalNumber(string: "\(WBPreferences.defaultTaxPercentage())")
        formatter.maximumFractionDigits = 2
        formatter.minimumFractionDigits = 2
        formatter.minimumIntegerDigits = 1
        
        priceSubject.subscribe(onNext: { [unowned self] value in
            if let val = value {
                let price = NSDecimalNumber(decimal: val)
                let tax = TaxCalculator.taxWithPrice(price, taxPercentage: self.taxPercentage, preTax: self.priceIsPreTax)
                self.taxSubject.onNext(self.formatter.string(from: tax)!)
            } else {
                self.taxSubject.onNext("")
            }
        }).disposed(by: disposeBag)
    }
    
    class func taxWithPrice(_ price: NSDecimalNumber, taxPercentage: NSDecimalNumber, preTax: Bool = false) -> NSDecimalNumber {
        var tax = NSDecimalNumber()
        if preTax {
            tax = price.multiplying(by: taxPercentage).dividing(by: HUNDRED)
        } else {
            let addeage = HUNDRED.adding(taxPercentage)
            let divided = price.dividing(by: addeage)
            tax = divided.multiplying(by: taxPercentage)
        }
        return tax
        
    }
}
